itests/docker/docker.go

```diff
@@ -10,239 +13,186 @@ import (
-		if err != nil {
-			return err
+	if c.errors != nil {
+		if err := c.errors.Close(); err != nil {
+			return errors.Wrap(err, "failed to close errors file")
 		}
 	}
-
 	return nil
 }
 
-func (d *Docker) RunContainers(ctx context.Context, paths config.ConfigPaths, suiteName string, goDesiredReward string,
-	goSupportedFeatures string) (*Ports, error) {
-	pwd, err := os.Getwd()
+type Docker struct {
+	suite string
+
+	pool    *dockertest.Pool
+	network *dockertest.Network
+
+	goNode    *NodeContainer
+	scalaNode *NodeContainer
+
+	logs string
+}
+
+// NewDocker creates a new Docker handler for a given suite name.
+// It removes any existing containers or networks for the suite and creates a new suite network.
+func NewDocker(suiteName string) (*Docker, error) {
+	pool, err := dockertest.NewPool("")
 	if err != nil {
 		return nil, err
 	}
-	err = os.MkdirAll(filepath.Join(pwd, logsDir, suiteName), os.ModePerm)
-	if err != nil {
-		return nil, err
+	docker := &Docker{suite: suiteName, pool: pool}
+	if rmErr := docker.removeContainers(); rmErr != nil {
+		return nil, rmErr
+	}
+	if rmErr := docker.removeNetworks(); rmErr != nil {
+		return nil, rmErr
 	}
+	if netErr := docker.createNetwork(); netErr != nil {
+		return nil, netErr
+	}
+	if mkErr := docker.mkLogsDir(); mkErr != nil {
+		return nil, mkErr
+	}
+	return docker, nil
+}
 
-	goNodeRes, goPorts, err := d.runGoNode(ctx, paths.GoConfigPath, suiteName, goDesiredReward, goSupportedFeatures)
+func (d *Docker) GoNode() *NodeContainer {
+	return d.goNode
+}
+
+func (d *Docker) ScalaNode() *NodeContainer {
+	return d.scalaNode
+}
+
+// StartGoNode starts a Go node container with the given configuration.
+func (d *Docker) StartGoNode(ctx context.Context, cfg config.DockerConfigurator) error {
+	var err error
+	d.goNode, err = d.startNode(ctx, cfg, goNodeLogFileName, goNodeErrFileName)
 	if err != nil {
-		return nil, err
+		return errors.Wrap(err, "failed to start Go node")
 	}
-	d.goNode = goNodeRes
-	scalaNodeRes, scalaPorts, err := d.runScalaNode(ctx, paths.ScalaConfigPath, suiteName)
+	return nil
+}
+
+// StartScalaNode starts a Scala node container with the given configuration.
+func (d *Docker) StartScalaNode(ctx context.Context, cfg config.DockerConfigurator) error {
+	var err error
+	d.scalaNode, err = d.startNode(ctx, cfg, scalaNodeLogFileName, scalaNodeErrFileName)
 	if err != nil {
-		return nil, err
+		return errors.Wrap(err, "failed to start Scala node")
 	}
-	d.scalaNode = scalaNodeRes
-
-	return &Ports{
-		Go:    goPorts,
-		Scala: scalaPorts,
-	}, nil
+	return nil
 }
 
 func (d *Docker) Finish(cancel context.CancelFunc) {
 	if d.scalaNode != nil {
 		err := d.pool.Client.KillContainer(dc.KillContainerOptions{
-			ID:     d.scalaNode.Container.ID,
+			ID:     d.scalaNode.container.Container.ID,
 			Signal: dc.SIGINT,
 		})
 		if err != nil {
-			log.Warnf("Failed to stop scala container: %v", err)
+			log.Printf("Failed to stop scala container: %v", err)
```

Comment:
Don't know why. Because of levels? It's for integration tests, nobody configure levels here. But slog text messages format not so nice because they try to keep the structure.

Reply:
Okay, fair enough